### ISSUE TITLE
[no-unnecessary-condition] `== null` against optional method param no longer allowed

### ISSUE THREAD

**Issue by leaumar** [typescript-eslint/typescript-eslint#2113]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error",
    "eqeqeq": ["error", "smart"],
  }
}
```

```TS
function foo(bar?: string) {
  if (bar != null) {
    // ^ "unnecessary condition"
  }
}
```

**Expected Result**
`== null` should be permitted against a `T?` value, especially when `eqeqeq` is set to `smart`.

**Actual Result**
A value that may be `undefined` cannot be `== null`.

**Rationale**
`== undefined` violates `eqeqeq=smart`. `===` against either one is less convenient and less safe at runtime where the distinction between them is rarely meaningful anyway. `== null` is the shortest to read/write and the most familiar idiom from other programming languages.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.0.1` |
| `@typescript-eslint/parser`        | `3.0.1` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `13.14.0` |
| `yarn`                              | `1.22.4` |
